gazel.go

```diff
@@ -217,58 +214,119 @@ func (v *Vendorer) updateSinglePkg(path string) error {
 	return v.updatePkg(path, "", pkg)
 }
 
+type RuleType int
+
+const (
+	RuleTypeGoBinary RuleType = iota
+	RuleTypeGoLibrary
+	RuleTypeGoTest
+	RuleTypeGoXTest
+	RuleTypeCGoGenrule
+)
+
+func (rt RuleType) RuleKind() string {
+	switch rt {
+	case RuleTypeGoBinary:
+		return "go_binary"
+	case RuleTypeGoLibrary:
+		return "go_library"
+	case RuleTypeGoTest:
+		return "go_test"
+	case RuleTypeGoXTest:
+		return "go_test"
+	case RuleTypeCGoGenrule:
+		return "cgo_genrule"
+	}
+	panic("unreachable")
+}
+
+type NamerFunc func(RuleType) string
+
 func (v *Vendorer) updatePkg(path, _ string, pkg *build.Package) error {
-	var rules []*bzl.Rule
 
-	var attrs Attrs = make(Attrs)
 	srcs := asExpr(merge(pkg.GoFiles, pkg.SFiles)).(*bzl.ListExpr)
+	cgoSrcs := asExpr(merge(pkg.CgoFiles, pkg.CFiles, pkg.CXXFiles, pkg.HFiles)).(*bzl.ListExpr)
 
 	deps := v.extractDeps(pkg.Imports)
 
-	if len(srcs.List) == 0 {
+	rules := v.emit(srcs, cgoSrcs, deps, pkg, func(rt RuleType) string {
+		switch rt {
+		case RuleTypeGoBinary:
+			return filepath.Base(pkg.Dir)
+		case RuleTypeGoLibrary:
+			return "go_default_library"
+		case RuleTypeGoTest:
+			return "go_default_test"
+		case RuleTypeGoXTest:
+			return "go_default_xtest"
+		case RuleTypeCGoGenrule:
+			return "cgo_codegen"
+		}
+		panic("unreachable")
+	})
+
+	wrote, err := ReconcileRules(filepath.Join(path, "BUILD"), rules, v.dryRun)
+	if err != nil {
+		return err
+	}
+	if wrote {
+		fmt.Fprintf(os.Stderr, "wrote BUILD for %q\n", pkg.Dir)
+	}
+	return nil
+}
+
+func (v *Vendorer) emit(srcs, cgoSrcs, deps *bzl.ListExpr, pkg *build.Package, namer NamerFunc) []*bzl.Rule {
+	var attrs Attrs = make(Attrs)
+	var rules []*bzl.Rule
+
+	if len(srcs.List) >= 0 {
+		attrs.Set("srcs", srcs)
+	} else if len(cgoSrcs.List) == 0 {
 		return nil
 	}
-	attrs.Set("srcs", srcs)
 
 	if len(deps.List) > 0 {
 		attrs.Set("deps", deps)
 	}
-
 	if pkg.IsCommand() {
-		rules = append(rules, newRule("go_binary", filepath.Base(pkg.Dir), attrs))
+		rules = append(rules, newRule(RuleTypeGoBinary, namer, attrs))
 	} else {
-		rules = append(rules, newRule("go_library", "go_default_library", attrs))
+		addGoDefaultLibrary := len(cgoSrcs.List) > 0 || len(srcs.List) > 0
+		if len(cgoSrcs.List) != 0 {
+			cgoRule := newRule(RuleTypeCGoGenrule, namer, map[string]bzl.Expr{
+				"srcs":      cgoSrcs,
+				"clinkopts": asExpr([]string{"-lz", "-lm", "-lpthread", "-ldl"}),
+			})
+			rules = append(rules, cgoRule)
+			attrs["library"] = asExpr(namer(RuleTypeCGoGenrule))
+		}
 		if len(pkg.TestGoFiles) != 0 {
-			rules = append(rules, newRule("go_test", "go_default_test", map[string]bzl.Expr{
-				"srcs":    asExpr(pkg.TestGoFiles),
-				"deps":    v.extractDeps(pkg.TestImports),
-				"library": asExpr("go_default_library"),
-			}))
+			xtestRule := newRule(RuleTypeGoTest, namer, map[string]bzl.Expr{
```

Comment:
xtests and regular tests both use RuleTypeGoTest (they are both just go_test targets).